### DISCUSSION TITLE
Problems writing SCP for amiga copy protected games

### DISCUSSION THREAD

**Discussion by RetroD** [keirf/greaseweazle#469] [Q&A] [answered]
Hello, I have briefly mentioned this in another thread, I am having a hard time successfully writing back a .scp image of a copy protected amiga game.

This is the summary:
- I can read to scp fine, and the images run in WinUAE.
- I can convert these images to HFEv3 in HxC and they work on a gotek drive in a real Amiga500
- I can WRITE IPF images I have downloaded and the new floppies read well on a real Amiga
- I can read and write back adf images of cracked games and they read fine on my Amiga

I have tried this on the following original amiga games I own:
- Mortal Kombat
- Project-X revised edition
- Sword of Sodan

However, no matter what I have tried, I cannot run games on floppies I have created by writing my SCP files back to floppy.
I have searched the web for the last 2-3 nights but I am unable to find a solution, or even confirm whether it is possible or not to do what I am trying...

I am using the greaseweasle 4.1 (sold by amigakit) and version 1.18 of the tools on windows 10.

From all my tests I believe the floppies and the drives I am using are good.
Is there anything I can try to successfully write SCP of copy protected games, or is it a fools errand?

**Comment by keirf** [comment, marked as answer]
The problem is that unanalysed SCP dumps cannot be intelligently written back to disk. These are the limitations:
1. No verification of the write
2. The write splice is always at the index (ie. the write is always index-to-index).
This is because `gw` doesn't know where the data is in the track. Some tools attempt to detect the write splice, and I think that might work for your Mortal Kombat dump, but `gw` doesn't attempt this.

To see what I mean, install HxC tools and inspect your SCP dump in the track analyzer. If you enter "disk view" you will see little red speckles all around. These are the write splices. Note how they are not forming a horizontal radial line at the index!

So, loads of sectors are getting corrupted by placing the write splice at the index.

I don't know about your Project-X and Sword of Sodan dumps, but the problem could well be the same.

Here is a method to generate IPFs from your SCP dumps.

Download my disk-utilities project. Either build for Linux/Mac from https://github.com/keirf/disk-utilities, or download a Windows prebuilt from this EAB thread: https://eab.abime.net/showthread.php?t=62217&page=51 (look for latest post from user acd2001).

Once installed (that's the battle, except the Windows prebuilt should be easy), then for example:
```
disk-analyse --format="Mortal Kombat" MK_1.scp MK_1.ipf
``` 

Wow! Now you have your own IPF from your own dump :) You can write this to disk or convert to HFEv3 using `gw`.

**Comment by keirf** [comment]
By the way, relating to #468, I doubt the HFEv3 you generated for Mortal Kombat using HxC is actually 100%, because of the non-index-aligned tracks. I tried
```
gw convert MK_1_scp.hfe tmp.scp
disk-analyse --format="Mortal Kombat" tmp.scp tmp.ipf
```
And there were loads of bad sectors. This is because the analyser is losing sync at the index mark, because the sector there is corrupt.

Anyway, now you have some things to look for, like non-index-alignment, and new tools to try, like disk-utilities:disk-analyse.

If you find a SCP image that *does* look index aligned (that should still be most commercial software), and want it handled for conversion to HFEv3 by `gw`, let me know. I can still add support for that, if/when it's useful.

**Comment by RetroD** [comment]
Thank you @keirf! This is all great information.
I knew about the disk utilities but I knew I would struggle to get the code compiled. I did not know a compiled version for windows existed on the web! (not that I searched extensively yet)

I spent sometime testing SCP to IPF conversion and then writing to a real floppy... all 3 disks I was testing worked!
Are there any plans to include that utility to the Greaseweazle program? It seems like it would be a perfect combination

Is the IPF conversion tool relying on officially created IPF files structure to write images? Meaning it is only able to convert to IPF if such IPF exists. I see it needs information from the format file to work correctly.

I will also be experimenting with other original disks I have. I am curious if index aligned tracks look visually different in  the track analyzer tool. I have noticed some ADF files look like pie slices, I wonder if that's what it is...

**Reply by keirf** [reply]
Yeah, I should make Windows automated builds on GitHub, really. Just need a bit of time to set that up. But acd2001 is very good at putting out new builds manually.

I don't know about integrating into gw. It sort of makes sense, it would allow direct integration into read and write commands. It's a bunch of work though.

The IPF creation is entirely independent of any SPS code or images. The IPF is generated totally fresh from your raw dump plus disk-analyse's understanding of 100s of Amiga disk formats. You can find a complete list here (and many other titles just haven't been catalogued yet): https://github.com/keirf/disk-utilities/blob/master/disk-analyse/formats

Index-aligned disks look like every track starts at the same offset, and write splice (often a red mark in the visualiser) will always be in same place, forming a horizontal radial line in the disk view in HxC. And indeed it will also give you the "pie" appearance for disks where HxC understands the sector format and paints them green. A disk formatted on Amiga will look like a random mess as all the tracks start at different positions around the disk.

**Reply by RetroD** [reply]
Yeah I can imagine it would not be trivial to integrate disk analyse , and you can still do everything you need while it's standalone, not a lot of gain for the work needed I guess.

So if I have a dump which is not on the format list, such as Lionheart, would I have to guess/experiment with the correct format by looking at others and add it to the list, or is there a more "scientific" way to determine that?

**Reply by keirf** [reply]
Formats `probe_all` and `probe_amiga` can be useful for applying all (or many common) analysers to each track to look for matches. You will get false positives though, especially as Amiga custom formats often have very weak (or even no!) checksum.

The best recommendation I can give you is to ask questions on the English Amiga Board thread on which the prebuilts get posted. That's the de facto support thread for disk-utilities. There are experienced users there who can help identify a new format, or tell you definitively if a title isn't supported yet. And the current most prolific author of new analysers (Abaddon) is also on there and you might convince him to support a major title like Lionheart if it isn't yet!

**Reply by RetroD** [reply]
Sounds good, thank you!
